backend-rs/src/utils.rs

```diff
@@ -0,0 +1,425 @@
+}
+
+impl CameraThreadHandle {
+    fn start_camera_thread(
+        camera_path: &str,
+        camera_handles: Arc<Mutex<Vec<CameraThreadHandle>>>,
+    ) -> Result<Self, Box<dyn Error>> {
+        let camera_path = camera_path.to_owned();
+        let (cam_mode_tx, mut cam_mode_rx) = mpsc::channel::<CameraMode>(1);
+        let manual_shutdown_needed: Arc<AtomicBool> = Arc::new(AtomicBool::new(false));
+        let sink_flush_needed: Arc<AtomicBool> = Arc::new(AtomicBool::new(false));
+        let tx_sink: Arc<Mutex<Vec<Sender<Vec<u8>>>>> = Arc::new(Mutex::new(Vec::new()));
+
+        let c_camera_path = camera_path.clone();
+        let c_manual_shutdown_needed = manual_shutdown_needed.clone();
+        let c_sink_flush_needed = sink_flush_needed.clone();
+        let c_tx_sink = tx_sink.clone();
+
+        let device_path = Path::new(&c_camera_path);
+        let node = v4l::context::enum_devices()
+            .into_iter()
+            .find(|node| node.path() == device_path)
+            .ok_or("V4l Node Not Found")?;
+        let mut device = Device::new(node.index())?;
+        let modes = CameraMode::fetch_all(&device)?;
+        let initial_mode = *modes.last().ok_or("No Last")?; // The last camera mode tends to be the one with the best resolution and fps.
+        thread::spawn(move || {
+            let mut reader = H264CameraReader::new(&mut device, initial_mode).unwrap();
+            let mut rtc_txs: Vec<Sender<Vec<u8>>> = Vec::new();
+
+            loop {
+                // Shutdown due to drop of CameraThreadHandle
+                if c_manual_shutdown_needed.load(Ordering::SeqCst) {
+                    return;
+                }
+
+                // Used to prevent having to constantly lock the c_tx_sink mutex.
+                if c_sink_flush_needed.load(Ordering::SeqCst) {
+                    let tx_sink = &mut *c_tx_sink.blocking_lock();
+
+                    while let Some(tx) = tx_sink.pop() {
+                        rtc_txs.push(tx);
+                    }
+
+                    c_sink_flush_needed.store(false, Ordering::SeqCst);
+                }
+
+                // Restart reader / stream with the new mode
+                if let Ok(mode) = cam_mode_rx.try_recv() {
+                    drop(reader);
+                    reader = H264CameraReader::new(&mut device, mode).unwrap();
+                }
+
+                let bytes = reader.read().unwrap();
+                rtc_txs.retain(|tx| tx.blocking_send(bytes.clone()).is_ok());
+
+                if rtc_txs.is_empty() {
+                    let camera_handles = &mut *camera_handles.blocking_lock();
+                    camera_handles.retain(|handle| handle.camera_path != c_camera_path);
+
+                    return;
+                }
+            }
+        });
+
+        Ok(CameraThreadHandle {
+            camera_path,
+            manual_shutdown_needed,
+            cam_mode_tx,
+            current_mode: *modes.last().ok_or("No Last Mode")?,
```

Comment:
Be careful with errors like this! These can lack context at competition. 

In a future PR, I'll add `tracing` to the project to help better log the camera/server state. For now, you might want to add something like "Error creating camera thread: failed to initialize camera". (this also goes for the other usage ~20 lines up)